install/scripts/craycli_init.py

```diff
@@ -0,0 +1,753 @@
+
+    # check for copy failure
+    outStr = cpOut.stdout.decode().rstrip()
+    rc = cpOut.returncode
+    if rc == 0:
+        # copying the script was successful - run the script command on the
+        # remote host using ssh - capture output
+        cmdStr = f"python3 {REMOTE_FILE_NAME} {cmdOpt}"
+        exeOut = subprocess.run(["ssh","-oStrictHostKeyChecking=no", host, cmdStr], 
+            shell=False, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
+
+        # if the script file is not present on the remote machine, we need to
+        # interpret that and intercept the error message here
+        outStr = exeOut.stdout.decode().rstrip()
+        rc = exeOut.returncode
+        if rc == 1 and "python3: can't open file" in outStr:
+            rc = MSG_MISSING_SCRIPT
+        elif rc == 1 and "Traceback" in outStr:
+            rc = MSG_PYTHON_SCRIPT_ERROR
+    else:
+        # failed to copy the script to the remote host
+        rc = MSG_REMOTE_COPY_FAILURE
+
+    # report results for this host in a thread safe manner
+    report_remote_init_results(host, rc, outStr)
+
+# mechanism for reporting remote initialization results
+def report_remote_init_results(host, retCode, message):
+    # add the item to the results list
+    global remote_init_results
+    global rir_mutex
+    with rir_mutex:
+        remote_init_results[host] = (retCode, message)
+
+# get the access token to use the k8s gateway
+def getToken(k8sClientApi):
+    # get the secret to request a token for gateway access
+    LOGGER.debug(f"Getting k8s gateway access token")
+    cs = ""
+    try:
+        sec = k8sClientApi.read_namespaced_secret("admin-client-auth","services").data
+        cs = base64.b64decode(sec["client-secret"])
+        LOGGER.debug(f"Read client auth secret")
+    except client.exceptions.ApiException as err:
+        LOGGER.error(f"Unable to read keycloak client auth secret: {err}")
+        raise SystemExit(err)
+
+    # construct the http call information
+    keycloak_endpoint = DEFAULT_KEYCLOAK_BASE + "/realms/shasta/protocol/openid-connect/token"
+    req_body = {
+        'grant_type': 'client_credentials',
+        'client_id': 'admin-client',
+        'client_secret': cs 
+    }
+
+    # make the call and parse response
+    token = ""
+    try:
+        LOGGER.debug(f"Querying keycloak for access token")
+        response = requests.post(keycloak_endpoint, data=req_body)
+        response.raise_for_status()
+        token = response.json()["access_token"]
+    except requests.exceptions.ConnectionError as err:
+        LOGGER.error(f"Unable to connect to keycloak endpoint: {err}")
+        raise SystemExit(err)
+    except requests.exceptions.HTTPError as err:
+        LOGGER.error(f"Unable to obtain gateway token: {err}")
+        raise SystemExit(err)
+
+    return token
+
+# get the management nodes from sls
+def getSlsNodes(k8sClientApi):
+    # TODO: with powerDNS changes, need to add '.hmn' or '.nmn' to the hostnames
```

Comment:
File a Jira that captures this work so it is not lost. Reference the Jira here in the code.

Reply:
Done.